### DISCUSSION TITLE
Prisma.validator

### DISCUSSION THREAD

**Discussion by amardeep911** [prisma/prisma#20658] [Q&A]
### Question

```js
export const participantPopulated =
  Prisma.validator<Prisma.ConversationParticipantInclude>()({
    user: {
      select: {
        id: true,
        username: true,
      },
    },
  });
<img width="888" alt="Screenshot 2023-08-11 at 7 47 37 PM" src="https://github.com/prisma/prisma/assets/97588898/94e494c8-77ad-4109-8924-aab8ea3ad4ed">
```
why is it saying "Property 'validator' does not exist on type 'typeof Prisma'.ts(2339)
any"

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @amardeep911 👋 

Can you please show how you are importing the `Prisma` object? Also, can you provide more information about the code where you use the `validator` method so we can further assist you?

**Reply by amardeep911** [reply]
import { GraphQLError } from 'graphql';
import { withFilter } from 'graphql-subscriptions';
import { Prisma } from 'prisma/prisma-client';
import { ConversationPopulated, GraphQLContext } from '../../../util/type';

interface Participant {
  userId: string;
  // Other participant properties
}

interface Conversation {
  participants: Participant[];
  // Other conversation properties
}
const resolvers = {
  Query: {
    conversations: async (
      _: any,
      __: any,
      context: GraphQLContext
    ): Promise<Array<ConversationPopulated>> => {
      const { session, prisma } = context;
      if (!session?.user) {
        throw new GraphQLError('You must be authenticated');
      }
      const {
        user: { id: userId },
      } = session;
      try {
        const conversations = await prisma.conversation.findMany({
          // It should be worked and it is a correct query confirmerd by the prisma team. Issue seems specific to Mongo
          // where: {
          //   participants: {
          //     some: {
          //       userId: {
          //         equals: userId,
          //       }
          //     },
          //   },
          // },
          include: conversationPopulated,
        });

        //Since that query is not working, we are doing this
        return conversations.filter(
          (conversation: Conversation) =>
            !!conversation.participants.find(
              participant => participant.userId === userId
            )
        );
      } catch (err: any) {
        console.log('conversation err', err);
        throw new GraphQLError(err?.message);
      }
    },
  },

  Mutation: {
    createConversation: async (
      _: any,
      args: { participantIds: Array<string> },
      context: GraphQLContext
    ): Promise<{ conversationId: string }> => {
      const { session, prisma, pubsub } = context;
      const { participantIds } = args;

      if (!session?.user) {
        throw new GraphQLError('You must be authenticated');
      }

      const {
        user: { id: userId },
      } = session;

      try {
        const conversation = await prisma.conversation.create({
          data: {
            participants: {
              createMany: {
                data: participantIds.map(participantId => ({
                  userId: participantId,
                  hasSeenLatestMessage: participantId === userId,
                })),
              },
            },
          },
          include: conversationPopulated,
        });

        pubsub.publish('CONVERSATION_CREATED', {
          conversationCreated: conversation,
        });

        console.log('conversation', conversation);

        return { conversationId: conversation.id };
      } catch (err) {
        console.log(err);
        throw new GraphQLError('Error creating conversation');
      }
    },
    markConversationAsRead: async function (
      _: any,
      args: { conversationId: string; userId: string },
      context: GraphQLContext
    ): Promise<boolean> {
      const { session, prisma } = context;
      const { conversationId, userId } = args;

      if (!session?.user) {
        throw new GraphQLError('You must be authenticated');
      }

      try {
        const participant = await prisma.conversationParticipant.findFirst({
          where: {
            userId,
            conversationId,
          },
        });

        if (!participant) {
          throw new GraphQLError(
            'You are not authorized to mark this conversation as read'
          );
        }

        await prisma.conversationParticipant.update({
          where: {
            id: participant.id,
          },
          data: {
            hasSeenLatestMessage: true,
          },
        });

        return true;
      } catch (error: any) {
        console.log('markConversationAsRead error', error);
        throw new GraphQLError(error?.message);
      }
    },
  },

  Subscription: {
    conversationCreated: {
      // subscribe: (_: any, __: any, context: GraphQLContext) => {
      //   const { pubsub } = context;
      //   return pubsub.asyncIterator('CONVERSATION_CREATED');
      // },
      subscribe: withFilter(
        (_: any, __: any, context: GraphQLContext) => {
          const { pubsub } = context;
          return pubsub.asyncIterator(['CONVERSATION_CREATED']);
        },
        (
          payload: ConversationCreatedSubscriptionPayload,
          _,
          context: GraphQLContext
        ) => {
          const { session } = context;
          const {
            conversationCreated: { participants },
          } = payload;

          console.log('payload', payload);

          const userIsParticipant = !!participants.find(
            (p: any) => p.user.id === session?.user?.id
          );
          return userIsParticipant;
        }
      ),
    },
  },
};

export interface ConversationCreatedSubscriptionPayload {
  conversationCreated: ConversationPopulated;
}

export const participantPopulated =
  Prisma.validator<Prisma.ConversationParticipantInclude>()({
    user: {
      select: {
        id: true,
        username: true,
      },
    },
  });

export const conversationPopulated =
  Prisma.validator<Prisma.ConversationInclude>()({
    participants: {
      include: participantPopulated,
    },
    latestMessage: {
      include: {
        sender: {
          select: {
            id: true,
            username: true,
          },
        },
      },
    },
  });

export default resolvers;

**Comment by MarkShawn2020** [comment]
same here

**Comment by tybro0103** [comment]
Not sure if this is OP's problem, but one little gotcha is `.validator` exists on `Prisma` from `'@prisma/client'` and not on your client instance.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
